From 6a6b16df28a0a38ac93a51767fcd743d04097ac5 Mon Sep 17 00:00:00 2001
From: Simon Marlow <marlowsd@gmail.com>
Date: Wed, 13 Apr 2011 11:27:20 +0100
Subject: [PATCH] fix a bug introduced in
 1fb38442d3a55ac92795aa6c5ed4df82011df724, symptom was 2047(threaded2) was
 crashing.

---
 rts/sm/GC.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

Index: ghc-6.12.3/rts/sm/GC.c
===================================================================
--- ghc-6.12.3.orig/rts/sm/GC.c
+++ ghc-6.12.3/rts/sm/GC.c
@@ -738,8 +738,12 @@ SET_GCT(gc_threads[0]);
   // zero the scavenged static object list 
   if (major_gc) {
       nat i;
-      for (i = 0; i < n_gc_threads; i++) {
-          zero_static_object_list(gc_threads[i]->scavenged_static_objects);
+      if (n_gc_threads == 1) {
+          zero_static_object_list(gct->scavenged_static_objects);
+      } else {
+          for (i = 0; i < n_gc_threads; i++) {
+              zero_static_object_list(gc_threads[i]->scavenged_static_objects);
+          }
       }
   }
 
